db.go

```diff
@@ -660,6 +666,31 @@ func (a *dbAppender) Rollback() error {
 	return g.Wait()
 }
 
+// Delete implements deletion of metrics.
+func (db *DB) Delete(mint, maxt int64, ms ...labels.Matcher) error {
+	db.cmtx.Lock()
+	defer db.cmtx.Unlock()
```

Comment:
As you said, another mutex is not exactly pretty. An alternative may be adding a function channel for coordinating block tasks.

Let's say there was `db.actorc` of type `chan func()`. The DB's backgronud routine would consume these functions in a for loop and execute them. That guarantees synchronization without the mutex. Additionally, if we decide to go the context route in #84  at some point, one can easily time out waiting for ones turns without having `Delete` block forever until it finally gets the mutex.

```go

func (db *DB) delete(mint, maxt int64, ms ...labels.Matcher) error {
    // actual delete logic.
}

func (db *DB) Delete(ctx context.Context, mint, maxt int64, ms ...labels.Matcher) (err error_ {
    select {
    case <-ctx.Done():
        err = ctx.Err()
    case db.actorc <- func() {
        err = db.delete(mint, maxt, ms...)
    }
    return err
}
```